include/verilated_timing.h

```diff
@@ -0,0 +1,367 @@
+        return range.first != range.second;
+    }
+
+    // Insert a coroutine waiting on the specified events
+    void insert(const VerilatedEventSet& events, std::coroutine_handle<> coro) {
+        for (auto event : events) m_eventsToEventSets.insert(std::make_pair(event, events));
+        m_eventSetsToCoros.insert(std::make_pair(events, coro));
+    }
+
+    // Erase all coroutines waiting on the given event
+    void erase(VerilatedEvent* event) {
+        auto range = m_eventsToEventSets.equal_range(event);
+        for (auto it = range.first; it != range.second; ++it) m_eventSetsToCoros.erase(it->second);
+    }
+};
+
+//=============================================================================
+// VerilatedEventDispatcher
+// Object that manages event variables and suspended coroutines waiting on those event vars.
+// When an event gets triggered, the corresponding coroutine isn't resumed immediately. This is to
+// allow for all processes to get to their timing controls, so that order of execution of processes
+// doesn't matter as much. Coroutines are resumed later by the 'resumeTriggered` function. The only
+// exception to this is when an event gets triggered twice – then all coroutines up to the first
+// trigger get resumed. This is neccessary so we don't lose track of events getting triggered and
+// the order they're triggered in.
+class VerilatedEventDispatcher final {
+private:
+    // MEMBERS
+    VerilatedEventToCoroMap m_eventsToCoros;  // Mapping of events to coroutines
+    std::vector<VerilatedEvent*> m_triggeredEvents;  // Events triggered in the current time slot
+    VerilatedEventSet m_eventsToReset;  // Events to be reset at the start of a time slot
+    VerilatedCoroutineArray
+        m_readyCoros;  // Coroutines ready for resumption in the current time slot
+    VerilatedMutex m_mutex;  // Protects m_eventSetsToCoros, m_triggeredEvents, m_eventsToCoros,
+                             // and m_readyCoros
+
+    // METHODS
+    // Move coroutines waiting on events from m_triggeredEvents to m_readyCoros
+    void readyTriggered() {
+        VerilatedLockGuard guard{m_mutex};
+        std::vector<VerilatedEvent*> queue = std::move(m_triggeredEvents);
+        for (auto event : queue) m_eventsToCoros.move(event, m_readyCoros);
+    }
+
+    // Are there no ready coroutines?
+    bool readyEmpty() {
+        const VerilatedLockGuard guard{m_mutex};
+        return m_readyCoros.empty();
+    }
+
+    // Are there coroutines waiting on this set of events? (wrapper for thread safety)
+    bool isSetWaitedOn(const VerilatedEventSet& events) {
+        const VerilatedLockGuard guard{m_mutex};
+        return m_eventsToCoros.contains(events);
+    }
+
+public:
+    // Insert the specified coroutine waiting on the given set of events
+    void insert(const VerilatedEventSet& events, std::coroutine_handle<> coro) {
+        if (isSetWaitedOn(events)) { readyTriggered(); }
+        VerilatedLockGuard guard{m_mutex};
```

Comment:
This locks twice, better to have an isSetWaitedOnLocked that is private and call it after the LockGuard.  Also use isSetWaitedOnLocked inside isSetWaitedOn after calling VerilatedLockGuard.